src/shogun/kernel/CachedKernel.h

```diff
@@ -0,0 +1,241 @@
+		void
+		resize_kernel_cache(KERNELCACHE_IDX size, bool regression_hack = false);
+
+		/** set the lru time
+		 *
+		 * @param t the time to use
+		 */
+		inline void set_time(int32_t t)
+		{
+			kernel_cache.time = t;
+		}
+
+		/** update lru time of item at given index to avoid removal from cache
+		 *
+		 * @param cacheidx index in cache
+		 * @return if updating was successful
+		 */
+		inline int32_t kernel_cache_touch(int32_t cacheidx)
+		{
+			if (kernel_cache.index[cacheidx] != -1)
+			{
+				kernel_cache.lru[kernel_cache.index[cacheidx]] =
+				    kernel_cache.time;
+				return (1);
+			}
+			return (0);
+		}
+
+		/** check if row at given index is cached
+		 *
+		 * @param cacheidx index in cache
+		 * @return if row at given index is cached
+		 */
+		inline int32_t kernel_cache_check(int32_t cacheidx)
+		{
+			return (kernel_cache.index[cacheidx] >= 0);
+		}
+
+		/** check if there is room for one more row in kernel cache
+		 *
+		 * @return if there is room for one more row in kernel cache
+		 */
+		inline int32_t kernel_cache_space_available()
+		{
+			return (kernel_cache.elems < kernel_cache.max_elems);
+		}
+
+		/** initialize kernel cache
+		 *
+		 * @param size size to initialize to
+		 * @param regression_hack if hack for regression shall be applied
+		 */
+		void kernel_cache_init(int32_t size, bool regression_hack = false);
+
+		/** cleanup kernel cache */
+		void kernel_cache_cleanup();
+
+	protected:
+		/** needs to be overridden to copy data */
+		virtual void copy_data(const SGReferencedData& orig);
+
+		/** needs to be overridden to initialize empty data */
+		virtual void init_data();
+
+		/** needs to be overridden to free data */
+		virtual void free_data();
+
+	private:
+#ifndef DOXYGEN_SHOULD_SKIP_THIS
+		/**@ cache kernel evalutations to improve speed */
+		struct KERNEL_CACHE
+		{
+			/** index */
+			int32_t* index;
+			/** inverse index */
+			int32_t* invindex;
+			/** active2totdoc */
+			int32_t* active2totdoc;
+			/** totdoc2active */
+			int32_t* totdoc2active;
+			/** least recently used */
+			int32_t* lru;
+			/** occu */
+			int32_t* occu;
+			/** elements */
+			int32_t elems;
+			/** max elements */
+			int32_t max_elems;
+			/** time */
+			int32_t time;
+			/** active num */
+			int32_t activenum;
+
+			/** buffer */
+			KERNELCACHE_ELEM* buffer;
+			/** buffer size */
+			KERNELCACHE_IDX buffsize;
+		};
+
+		/** kernel thread parameters */
+		struct S_KTHREAD_PARAM
+		{
+			/** kernel */
+			CKernel* kernel;
+			/** kernel cache */
+			KERNEL_CACHE* kernel_cache;
+			/** cache */
+			KERNELCACHE_ELEM** cache;
+			/** uncached rows */
+			int32_t* uncached_rows;
+			/** number of uncached rows */
+			int32_t num_uncached;
+			/** needs computation */
+			uint8_t* needs_computation;
+			/** start */
+			int32_t start;
+			/** end */
+			int32_t end;
+			/** of vectors */
+			int32_t num_vectors;
+		};
+#endif // DOXYGEN_SHOULD_SKIP_THIS
+
+		static void* cache_multiple_kernel_row_helper(void* p);
```

Comment:
I assume all this is basically cut/pasted from somewhere else?

Reply:
yes, from Kernel.h